File: apps/tiddly_wiki_bot/lib/tiddly_wiki/tiddly_wiki.ex
defmodule TiddlyWiki do
  defstruct [
    :url,
    :external_url,
    :username,
    :password
  ]

  @type t :: %TiddlyWiki{
          url: String.t(),
          external_url: String.t(),
          username: String.t(),
          password: String.t()
        }

  @callback get_all(dest :: TiddlyWiki.t(), filter :: String.t()) ::
              {:ok, list(TiddlyWiki.Tiddler.t())}
  @callback get(dest :: TiddlyWiki.t(), title :: String.t()) :: {:ok, TiddlyWiki.Tiddler.t()}
  @callback put(dest :: TiddlyWiki.t(), tiddler :: TiddlyWiki.Tiddler.t()) :: {:ok, String.t()}

  defp from_env do
    TiddlyWiki
    |> struct(Application.get_env(:tiddlywiki_bot, :wiki))
  end

  def put(%TiddlyWiki.Tiddler{} = tiddler) do
    from_env()
    |> implementation().put(tiddler)
  end

  def get(title) do
    from_env()
    |> implementation().get(title)
  end

  def get_all(filter) do
    from_env()
    |> implementation().get_all(filter)
  end

  def get_all do
    from_env()
    |> implementation().get_all()
  end

  def absolute_url(tiddler) do
    "#{from_env().external_url}/#{tiddler.title}"
  end

  defp implementation,
    do:
      Application.get_env(
        :tiddlywiki_bot,
        :tiddlywiki_client
      )
end

File: apps/pixelfed2wiki/tiddlywiki/tiddlywiki.go
package tiddlywiki

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"os"

	"github.com/carlo-colombo/pixelfed2wiki/uploader"
	"github.com/minio/minio-go/v7"
)

type TiddlywikiClient struct {
	client   http.Client
	dest     string
	username string
	password string
}

func NewClient(dest string, username string, password string) TiddlywikiClient {
	return TiddlywikiClient{
		dest:     dest,
		client:   http.Client{},
		username: username,
		password: password,
	}
}

func (tc TiddlywikiClient) decorateReq(req *http.Request) *http.Request {
	req.Header.Add("x-requested-with", "TiddlyWiki")
	if tc.username != "" && tc.password != "" {
		req.SetBasicAuth(tc.username, tc.password)
	}
	return req
}

func (tc TiddlywikiClient) CreateIfNew(mc *minio.Client, tiddler Tiddler) error {
	path := fmt.Sprintf("%s/recipes/default/tiddlers/%s",
		tc.dest,
		url.PathEscape(tiddler.Title))

	req, err := http.NewRequest("GET", path, nil)
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req = tc.decorateReq(req)

	resp, err := tc.client.Do(req)

	if err != nil || resp.StatusCode != 404 {
		fmt.Printf("(%d)skipping creation %s \n", resp.StatusCode, tiddler.Title)
		return nil
	}

	uploader := uploader.NewUploader(
		os.Getenv("UPLOAD_BUCKET"),
		mc,
		tiddler.imageBytes,
	)

	if uploader.Err != nil {
		return fmt.Errorf("failed to instantiate uploader: %w", uploader.Err)
	}

	tiddler = tiddler.
		UploadAndSetImage(uploader).
		UploadAndSetThumbnail(uploader)

	if tiddler.Err != nil {
		return fmt.Errorf("failed to finalize tiddler: %w", tiddler.Err)
	}

	tjson, err := json.Marshal(tiddler)

	if err != nil {
		return fmt.Errorf("failed to marshal tiddler: %w", err)
	}

	req, err = http.NewRequest("PUT", path, bytes.NewBuffer(tjson))

	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req = tc.decorateReq(req)

	resp, err = tc.client.Do(req)

	if err != nil {
		return fmt.Errorf("failed to create tiddler: %w", err)
	}

	if resp.StatusCode != 204 {
		return fmt.Errorf("failed to create tiddler: %v", resp.Status)
	}

	fmt.Printf("created %s\n", tiddler.Title)
	return nil
}
